### DISCUSSION TITLE
API design for optimization functions

### DISCUSSION THREAD

**Discussion by bwpriest** [LLNL/MuyGPyS#121] [Ideas]
As we near a 1.0 release, we need to settle on a long-term interface for optimization functions.
These functions include `MuyGPyS.optimize.optimize_from_tensors`, `MuyGPyS.optimize.sigma_sq.muygps_sigma_sq_optim`, and any future optimization functions. 

===== Current API =====

Optimization functions currently operate as pure functions that `deepcopy` passed `MuyGPS` objects and modify and return the copies, leaving the original `MuyGPS` object unchanged. Ergo, the skeleton of such functions is:
```
def optim(muygps: MuyGPS, *args, **kwargs) -> MuyGPS:
    ret = deepcopy(muygps)
    ...
    return ret
```

===== Outstanding problems =====

The current design made sense when `MuyGPS` objects were lightweight and I was trying to make the library fully functional, because `optim` above has no side effects and the cost of copying `muygps` was low. However, we are starting to introduce larger numbers of hyperparameters - `HeteroscedasticNoise` and `AnisotropicDistortion` will potentially introduce thousands or millions of parameters that we don't want to have to copy during optimization. Furthermore, it is not clear to me whether it is worthwhile to keep the optimization functions "pure", since it is not obvious whether anyone will want to hang onto the original (unoptimized) models. 

===== Possible Solutions =====

1. We could stick with the current pure function design philosophy and create a more elegant copy method. This is what issue #105 suggests. The main benefit here is that it will be low effort - once we have a method that only deepcopies the optimization parameters, it can be inserted in-place without affecting the rest of the code or API.
2. We could drop the pure function design philosophy specifically for optimization. This would result in modifying hyperparameters in-place. I think this is how most users expect optimization to work, so it might be more intuitive. The main concern is how it would affect the API. We can modify the existing signatures to `optim(muygps: MuyGPS, *args, **kwargs) -> None`, and assure that the `muygps` object is mutated by the function. We could also make `optim` a member of `MuyGPS`. I do not like the latter idea, primarily because we may have many optimization functions in the future and I don't want to pollute the class APIs with lots of functions.

===== Discuss =====

Please sound off in this thread so that we can converge on a solution for a stable optimization API.

**Comment by alecmdunton** [comment]
I vote that we migrate to an in-place modification paradigm (Option 2). I don't see a need to keep the optimization component of the library functional - there is no use case for unoptimized GP models I can think of either.

**Reply by bwpriest** [reply]
If we do this we'll need to update all of the associated documentation and tests. We'll also need to be careful that it conforms with the new ways in which functors get constructed. For example, `MuyGPS.eps`, `MuyGPS._mean_fn.eps`, `MuyGPS._var_fn.eps`, and `MuyGPS._fast_precompute_fn.eps` all refer to the same object. We will need to be careful to update `eps` via `MuyGPS.eps._set(...)` rather than `MuyGPS.eps = ...`, since the latter will not update the other references and will result in the functors' using a different value than expected when invoked.
